ex5/h_miyaji/main.py

```diff
@@ -0,0 +1,418 @@
+    )
+    return log_likelihood
+
+
+def plot_one_line(
+    dataset: np.ndarray,
+    title: str,
+    xlabel: str = "x1",
+    ylabel: str = "x2",
+    zlabel: str = "x3",
+) -> None:
+    """Plot only one graph.
+
+    Args:
+        dataset (np.ndarray): data of x, y (and z) (n, dim).
+        title (str): the graph title.
+        xlabel (str, optional): label name on x-axis. Defaults to "x1".
+        ylabel (str, optional): label name on y-axis. Defaults to "x2".
+        zlabel (str, optional): label name on z-axis. Defaults to "x3".
+
+    Returns:
+        None:
+    """
+    dim = len(dataset[0])  # dimension of the data
+
+    fig = plt.figure()
+
+    x1 = dataset[:, 0]
+    x2 = dataset[:, 1]
+
+    if dim == 1 or dim == 2:  # x-y
+        ax = fig.add_subplot(111)
+        ax.plot(x1, x2, color="b")
+
+    elif dim == 3:  # x-y-z
+        x3 = dataset[:, 2]
+        ax = fig.add_subplot(111, projection="3d")
+        ax.plot(x1, x2, x3, color="b")
+        ax.set_zlabel(zlabel)
+
+    ax.set_title(title)
+    ax.set_xlabel(xlabel)
+    ax.set_ylabel(ylabel)
+    ax.grid()
+    # plt.savefig("h_miyaji\\figs\\result3-2.png")
+    plt.show()
+
+    return None
+
+
+def plot_gmm(
+    mix_gauss: np.ndarray,
+    means: np.ndarray,
+    x_value: np.ndarray,
+    axis_range: list,
+    ax,
+):
+    """Plot GMM on a scatter plot.
+
+    Args:
+        mix_gauss (np.ndarray): the data of GMM (n, )
+        means (np.ndarray): mean for each cluster (k, dim).
+        x_value (np.ndarray): the data of x1 (and x2) axis (plot_num, dim).
+        axis_range (list): the range of graph axis ([xmin, xmax, ymin, ymax]).
+        ax (Axes): Axes object of the scatter plot.
+
+    Returns:
+        ax: Axes object of the graph.
+    """
+    # dimension of the data
+    if isinstance(means[0], np.ndarray):
+        dim = len(means[0])
+    else:
+        dim = 1
+    colors = ("b", "r")  # color of plot (blue, red)
+
+    if dim == 1:  # x
+        ax.plot(x_value, mix_gauss, label="GMM", color=colors[0])
+        ax.scatter(
+            means,
+            np.zeros((len(means))),
+            label="Centroids",
+            color=colors[1],
+            marker="x",
+            s=200,
+        )
+
+    elif dim == 2:  # x-y
+        contour = ax.contour(x_value[0], x_value[1], mix_gauss, cmap="magma")
+        ax.scatter(
+            means[:, 0],
+            means[:, 1],
+            label="Centroids",
+            color=colors[1],
+            marker="x",
+            s=200,
+        )
+
+        # create continuous colorbar
+        norm = matplotlib.colors.Normalize(contour.cvalues.min(), contour.cvalues.max())
+        sm = plt.cm.ScalarMappable(norm=norm, cmap=contour.cmap)
+        sm.set_array([])
+        plt.colorbar(sm, ticks=contour.levels, ax=ax, label="Probability density")
+
+    ax.legend()
+    ax.axis(axis_range)
+    # plt.savefig("h_miyaji\\figs\\result3-1.png")
```

Comment:
コメントアウトしてある箇所なので動作の上での影響はないですが，OSによってパスの区切り文字の扱い方が異なるので，ベタ書きするのは避けるとベターです． `os.path.join()`を使うと，OSの差分を吸収してくれるのでこちらを使うことをオススメします．
```suggestion
    import os
    # plt.savefig(os.path.join("h_miyaji", "figs", "result3-1.png"))
```